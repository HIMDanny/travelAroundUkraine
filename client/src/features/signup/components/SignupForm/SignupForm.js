/* eslint-disable no-unused-vars */
import React, { useState } from 'react';
import { Stepper, StepLabel, Step, Button, Box, CircularProgress } from '@mui/material';
import { Formik, Form } from 'formik';
import { LocalizationProvider } from '@mui/x-date-pickers';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns';
import { enAU } from 'date-fns/locale';

import { initialValues, validationSchema } from '../../data';
import { UserInfoForm, LoginInfoForm, SignupSuccess } from '..';
import postSignUpData from '../../../../api/postSignUpData';

const steps = ['User Info', 'Login Info'];

const SignupForm = () => {
  const [activeStep, setActiveStep] = useState(0);
  const [isModalOpen, setIsModalOpen] = useState(false);
  const currentValidationSchema = validationSchema[activeStep];
  const lastStep = steps.length - 1;

  const handleCloseModal = () => setIsModalOpen(false);

  const nextStep = () => {
    setActiveStep((prev) => (prev >= lastStep ? prev : prev + 1));
  };

  const previousStep = () => {
    setActiveStep((prev) => prev - 1);
  };

  const formSubmitHandler = async (values) => {
    nextStep();

    if (activeStep === lastStep) {
      const { isSuccess, error } = await postSignUpData(values);
      if (isSuccess) {
        setIsModalOpen(true);
      } else {
        // TODO: render error alert
        // eslint-disable-next-line no-console
        console.error(error); // ! for demonstration purpose
      }
    }
  };

  return (
    <>
      <Stepper activeStep={activeStep} alternativeLabel sx={{ mb: 3 }}>
        {steps.map((label) => (
          <Step key={label}>
            <StepLabel>{label}</StepLabel>
          </Step>
        ))}
      </Stepper>

      <Formik initialValues={initialValues} validationSchema={currentValidationSchema} onSubmit={formSubmitHandler}>
        {({ isSubmitting }) => (
          <LocalizationProvider dateAdapter={AdapterDateFns} adapterLocale={enAU}>
            <Form>
              {activeStep === 0 && <UserInfoForm />}
              {activeStep === 1 && <LoginInfoForm />}

              <Box sx={{ display: 'flex', justifyContent: 'center', columnGap: 3, position: 'relative' }}>
                {activeStep !== 0 && !isSubmitting && <Button onClick={previousStep}>Back</Button>}

                {!isSubmitting ? (
                  <Button disabled={isSubmitting} type="submit">
                    {activeStep !== lastStep ? 'Continue' : 'Sign up'}
                  </Button>
                ) : (
                  <CircularProgress
                    size={50}
                    sx={{
                      color: 'primary.dark',
                    }}
                  />
                )}
              </Box>
            </Form>
          </LocalizationProvider>
        )}
      </Formik>

      <SignupSuccess open={isModalOpen} onClose={handleCloseModal} />
    </>
  );
};

export default SignupForm;
