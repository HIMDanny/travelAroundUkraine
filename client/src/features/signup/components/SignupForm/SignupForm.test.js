/* eslint-disable react/prop-types */
import { render } from '@testing-library/react';
import SignupForm from './SignupForm';

jest.mock('@mui/material/Grid', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Typography', () => ({ children }) => <p>{children}</p>);
jest.mock('@mui/material/Stepper', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Step', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/StepLabel', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Button', () => ({ children }) => <button>{children}</button>);
jest.mock('@mui/material/Box', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/CircularProgress', () => ({ children }) => <div>{children}</div>);
jest.mock('../SignupSuccess/SignupSuccess', () => () => <div>SIGN UP SUCCESS</div>);
jest.mock('../UserInfoForm/UserInfoForm', () => () => <div>USER INFO FORM</div>);

describe('SignupForm snapshot test', () => {
  test('should SignupForm match snapshot', () => {
    const { asFragment } = render(<SignupForm />);
    expect(asFragment()).toMatchSnapshot();
  });
});
