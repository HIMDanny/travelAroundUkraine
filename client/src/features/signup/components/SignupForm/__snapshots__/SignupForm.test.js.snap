// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm snapshot test should SignupForm match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div>
        User Info
      </div>
    </div>
    <div>
      <div>
        Login Info
      </div>
    </div>
  </div>
  <form
    action="#"
  >
    <div>
      USER INFO FORM
    </div>
    <div>
      <button>
        Continue
      </button>
    </div>
  </form>
  <div>
    SIGN UP SUCCESS
  </div>
</DocumentFragment>
`;
